src/components/annotations/draw.js

```diff
@@ -340,8 +369,29 @@ function drawRaw(gd, options, index, subplotId, xa, ya) {
                     annSizeFromHeight * shiftFraction(0.5, options.yanchor);
 
                 if(tailRef === axRef) {
-                    posPx.tail = ax._offset + ax.r2p(arrowLength);
-                    // tail is data-referenced: autorange pads the text in px from the tail
+                    // In the case tailRefType is 'domain' or 'paper', the arrow's
+                    // position is set absolutely, which is consistent with how
+                    // it behaves when its position is set in data ('range')
+                    // coordinates.
+                    var tailRefType = Axes.getRefType(tailRef);
+                    if(tailRefType === 'domain') {
+                        if(axLetter === 'y') {
+                            arrowLength = 1 - arrowLength;
+                        }
```

Comment:
We could avoid second repetition by moving `arrowLength = 1 - arrowLength;` block(s) one step up.
```js
var isY = axLetter === 'y';
if(isY) arrowLength = 1 - arrowLength;
if(tailRefType === 'domain') {
  posPx.tail = ax._offset + ax._length * arrowLength;
} else if(tailRefType === 'paper') {
  posPx.tail = isY ? 
    gs.t + gs.h * arrowLength :
    gs.l + gs.w * arrowLength;
} ...

Reply:
Wouldn't this make `arrowLength` wrong in the case that `tailRefType` is not `domain` or `paper`?

Reply:
Good call. Resolving.